File: hackerrank/Algorithms/python/counting-valleys.py
#!/bin/python3

import math
import os
import random
import re
import sys

def countingValleys(steps, path):
  count = 0
  valley = 0
  for x in path:
    if x == "U": 
      count+=1
      if count==0: valley+=1
    else:
      count-=1
  return valley

if __name__ == '__main__':
  fptr = open(os.environ['OUTPUT_PATH'], 'w')

  steps = int(input().strip())

  path = input()

  result = countingValleys(steps, path)

  print(result)

  fptr.write(str(result) + '\n')

  fptr.close()

File: hackerrank/Algorithms/go/counting-valleys.go
package main

import (
  "bufio"
  "fmt"
  "io"
  "os"
  "strconv"
  "strings"
)

func countingValleys(steps int32, path string) int32 {
  _ = steps
  var count, valley int32
  for _, c := range path {
    if c == 'U' {
      count += 1
      if count == 0 { valley += 1 }
    } else { count -= 1}
  }
  return valley
}

func main() {
  reader := bufio.NewReaderSize(os.Stdin, 16 * 1024 * 1024)

  stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
  checkError(err)

  defer stdout.Close()

  writer := bufio.NewWriterSize(stdout, 16 * 1024 * 1024)

  stepsTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
  checkError(err)
  steps := int32(stepsTemp)

  path := readLine(reader)

  result := countingValleys(steps, path)

  fmt.Fprintf(writer, "%d\n", result)

  writer.Flush()
}

func readLine(reader *bufio.Reader) string {
  str, _, err := reader.ReadLine()
  if err == io.EOF { return "" }
  return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
  if err != nil { panic(err) }
}